### PR TITLE
HashCode documentation

### PR DESCRIPTION
**Pull request by jcdickinson**
## Summary

Documentation for `System.HashCode`.

I will add some comments surrounding some unknowns about how to document correctly.

Fixes dotnet/docs#4936

### CODE REVIEW

File: xml/System/HashCode.xml [hunk 1]
```diff
@@ -13,8 +13,54 @@
...
+single hash code using the xxHash32 algorithm. This structure operates in one of two ways:
+
+ * Static methods that accept a set of up to eight values to combine.
+ * Two instance methods that operate in a streaming fashion, accepting value one at a time.
+
+### Static Methods
+
+The static methods should be preferred when eight or fewer hash codes are being combined.
+ 
+ [!code-csharp[System.HashCode.Structure#1](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.hashcode.structure/cs/example1.cs#1)]
```
Comment (jcdickinson) [inline xml/System/HashCode.xml]: Where do I add these samples?

File: xml/System/HashCode.xml [hunk 2]
```diff
@@ -435,10 +479,9 @@
         <Parameter Name="obj" Type="System.Object" />
       </Parameters>
       <Docs>
-        <param name="obj">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="obj">Ignored.</param>
+        <summary>This method is not supported and should not be called.</summary>
+        <returns>This method will always throw a <see cref="System.NotSupportedException" />.</returns>
```
Comment (jcdickinson) [inline xml/System/HashCode.xml]: Not sure how to cref exceptions in this format.

File: xml/System/HashCode.xml [hunk 3]
```diff
@@ -94,11 +142,14 @@
         <Parameter Name="value1" Type="T1" />
       </Parameters>
       <Docs>
-        <typeparam name="T1">To be added.</typeparam>
-        <param name="value1">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <typeparam name="T1">The type of the value to mix.</typeparam>
```
Comment (jcdickinson) [inline xml/System/HashCode.xml]: NB: this overload is slightly different to the rest.

File: xml/System/HashCode.xml (hunk 1 above)
Comment (BillWagner) [reply xml/System/HashCode.xml]: See [the sample repository](https://github.com/dotnet/samples)

File: xml/System/HashCode.xml (hunk 2 above)
Comment (BillWagner) [reply xml/System/HashCode.xml]: Use the xref syntax:

```xml
<xref:System.NotSupportedException?displayProperty=nameWithType>
```

File: xml/System/HashCode.xml (hunk 1 above)
Comment (jcdickinson) [reply xml/System/HashCode.xml]: https://github.com/dotnet/samples/pull/92

**Comment by BillWagner** [comment]
closing and re-opening to trigger a new build

File: xml/System/HashCode.xml (hunk 2 above)
Comment (mairaw) [reply xml/System/HashCode.xml]: I think they looked correct before except they were missing the T:

Comment (jcdickinson) [reply xml/System/HashCode.xml]: Yep, I figured it out by grepping the existing docs.

Comment (mairaw) [reply xml/System/HashCode.xml]: Good. We'll review the PR soon. Thanks!

File: xml/System/HashCode.xml [hunk 4]
```diff
@@ -486,9 +532,13 @@
       </ReturnValue>
       <Parameters />
       <Docs>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <summary>
+          Calculates the final hash code after consecutive <see cref="System.HashCode.Add" /> invocations.
```
Comment (BillWagner) [inline xml/System/HashCode.xml]: cref should have "M:" prefixing the member name.

File: xml/System/HashCode.xml [hunk 5]
```diff
@@ -66,11 +110,16 @@
...
       <Docs>
-        <typeparam name="T">To be added.</typeparam>
-        <param name="value">To be added.</param>
-        <param name="comparer">To be added.</param>
-        <summary>To be added.</summary>
-        <remarks>To be added.</remarks>
+        <typeparam name="T">The type of the value to add to the hash code.</typeparam>
+        <param name="value">The value to add to the hash code.</param>
+        <param name="comparer">
+          The <see cref="System.Collections.Generic.IEqualityComparer&lt;T&gt;" /> to use to calculate the hash code.
```
Comment (BillWagner) [inline xml/System/HashCode.xml]: This cref should be prefaced with "T:"

File: xml/System/HashCode.xml [hunk 6]
```diff
@@ -13,8 +13,53 @@
...
+single hash code. This structure operates in one of two ways:
+
+ * Static methods that accept a set of up to eight values to combine.
+ * Two instance methods that operate in a streaming fashion, accepting values one at a time.
+
+<xref:System.HashCode> uses a static random seed. <xref:System.HashCode> returns stable values across a single process
+invocation, but not across multiple process invocations. Do not store hash codes produced by <xref:System.HashCode> in
+serialized structures, for example, on-disk.
+
+<xref:System.HashCode> currently returns the same value if <xref:System.HashCode.ToHashCode> is called multiple times.
```
Comment (BillWagner) [inline xml/System/HashCode.xml]: This paragraph is confusing to me. Without adding the detail you've added in the ToHashCode method, this implies that the same object could generate different hash codes. That sounds bad.

Going further, it implies that if a type supports value-based equality, two objects with the values (and therefore considered equal) may not always generate the same hash code. That also sounds bad.

I'm probably missing some important point, and that important point should be made more clear.

Comment (BillWagner) [reply xml/System/HashCode.xml]: I was worried when I read it that this test might fail:

```csharp
var hc1 = new HashCode();
hc1.Add(1);
hc1.Add(2);
var hashOne = hc1.ToHashCode();
var hc2 = new HashCode();
hc2.Add(1);
hc2.Add(2);
var hashTwo = hc2.ToHashCode();
Assert.Equals(hashOne, hashTwo);
```

I think I understand, but there's a bug in your test above (`b` isn't declared).  If you make a correction, I'll suggest different wording.

Comment (jcdickinson) [reply xml/System/HashCode.xml]: Yep, I meant to combine `(1, 2)`. I've submitted `Dispose`-inspired wording that simply omits the explanation as to why.

**Comment by jcdickinson** [comment]
Pleasure, looking forward to helping when I have the time again :)!

**Comment by Stefan137** [comment]
It is not clear to me from the documentation whether the order of _Add()_ calls is relevant or not (I suspect it is). For example, is the following test expected to succeed or to fail?

    var hc1 = new HashCode;
    hc1.Add(1);
    hc1.Add(2);
    var hc2 = new HashCode;
    hc2.Add(2);
    hc2.Add(1);
    Assert.Equals(hc1.ToHashCode(), hc2.ToHashCode());

Depending on the context (e.g., calculating the hash code of a set or of a list), it might be important to combine hash codes in order-sensitive manner or in order-insensitive manner.

**Pull request merged**